graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -578,6 +590,167 @@ def prune_output_blocks_in_compound_match_query(compound_match_query):
+                match_traversals=new_match_traversals,
+                folds=match_query.folds,
+                output_block=match_query.output_block
+            )
+        )
+
+    return CompoundMatchQuery(match_queries=new_match_queries)
+
+
+def _update_tagged_expression(expression, current_locations):
+    """Replace non-existent tag Expressons with True, and simplify the result."""
+    if isinstance(expression, BinaryComposition):
+        if isinstance(expression.left, ContextField):
+            context_field = expression.left
+        elif isinstance(expression.right, ContextField):
+            context_field = expression.right
+        elif expression.operator == u'||':
+            if expression.left == TrueLiteral or expression.right == TrueLiteral:
+                return TrueLiteral
+            else:
+                return expression
+        elif expression.operator == u'&&':
+            if expression.left == TrueLiteral and expression.right == TrueLiteral:
+                return TrueLiteral
+            if expression.left == TrueLiteral:
+                return expression.right
+            if expression.right == TrueLiteral:
+                return expression.left
+            else:
+                return expression
+        else:
+            return expression
+        location_name, _ = context_field.location.get_location_name()
+        if location_name not in current_locations:
+            return TrueLiteral
+        else:
+            return expression
+    elif isinstance(expression, TernaryConditional):
+        if expression.predicate == TrueLiteral:
+            return expression.if_true
+        else:
+            return expression
+    elif isinstance(expression, Filter):
+        if expression.predicate == TrueLiteral:
+            return None
+        else:
+            return expression
+    else:
+        return expression
+
+
+def _construct_update_tags_visitor_fn(current_locations):
+    """Return an Expression updater using the given `current_locations`."""
+    def visitor_fn(expression):
+        return _update_tagged_expression(expression, current_locations)
+    return visitor_fn
+
+
+def lower_filter_expressions_in_compound_match_query(compound_match_query):
+    """Replace Expressons involving non-existent tags with True."""
+    if len(compound_match_query.match_queries) == 1:
+        return compound_match_query
+    elif len(compound_match_query.match_queries) == 0:
+        raise AssertionError(u'Received CompoundMatchQuery with an empty list of MatchQueries.')
+    else:
+        new_match_queries = []
+        for match_query in compound_match_query.match_queries:
+            match_traversals = match_query.match_traversals
+            current_locations = set()
+
+            for traversal in match_traversals:
+                for step in traversal:
+                    if step.as_block is not None:
+                        location_name, _ = step.as_block.location.get_location_name()
```

Comment:
Any time you get 5 indentation levels deeper than where the function started, there is too much complexity and the function should be refactored and broken up. This is especially the case if 3 of the indentation levels are `for` loops.